### PR TITLE
Generate tsconfig once instead of on every webpack invocation

### PR DESCRIPTION
**Pull request by petemill**
Avoids any potential race condition with parallel webpack invocations

Fixes https://github.com/brave/brave-browser/issues/18416

What was happening was that multiple parallel processes were writing `gen/tsconfig.json`. Whilst they were all writing the exact same content, one or more of the processes would see a blank file immediately after writing.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/webpack/gen-tsconfig.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+const fs = require('fs-extra')
+const path = require('path')
+const pathMap = require('./path-map')
+
+const srcPath = path.resolve(__dirname, '../../../')
+const braveSrcPath = path.join(srcPath, 'brave')
+
+/**
+ * Generates a tsconfig.json file in the gen/ directoryc
```
Comment (bsclifton) [inline components/webpack/gen-tsconfig.js:9]: s/directoryc/directory

Comment (bsclifton) [inline components/webpack/gen-tsconfig.js:9]:
```suggestion
 * Generates a tsconfig.json file in the gen/ directory
```

File: components/webpack/webpack.config.js [hunk 2]
```diff
@@ -4,77 +4,15 @@
...
-    ]
-  }
-  await fs.writeFile(tsConfigPath, JSON.stringify(config))
-  return tsConfigPath
-}
+const GenerateDepfilePlugin = require('./webpack-plugin-depfile')
+const pathMap = require('./path-map')
 
 module.exports = async function (env, argv) {
   // TODO(petemill): only do this once per build, in a separate target
```
Comment (bsclifton) [inline components/webpack/webpack.config.js:13]: can we remove the TODO now? 🙂

Comment (petemill) [reply components/webpack/webpack.config.js:13]: 🙃

File: components/webpack/gen-tsconfig.js (hunk 1 above)
Comment (petemill) [reply components/webpack/gen-tsconfig.js]: fixed

File: components/webpack/webpack.config.js (hunk 2 above)
Comment (petemill) [reply components/webpack/webpack.config.js]: fixed

**Review by bsclifton — APPROVED** [review]
This is nice! 😄

**Pull request merged**